### ISSUE TITLE
formatter lsp: LSP formatting of special characters generates invalid results

### ISSUE THREAD

**Issue by nrayburn-tech** [oxc-project/oxc#23314]
Initially reported in https://github.com/oxc-project/oxc-intellij-plugin/issues/502, but I was able to confirm odd behavior with IntelliJ and VS Code. This does not apply with the Oxc playground or CLI, so it's pretty likely a language server issue or something odd the language server is doing when interacting with the formatter.

1. Clone https://github.com/nrayburn-tech/oxc-intellij-plugin-502
2. Run `npm install`
3. Trigger formatting through the language server.
4. Content output is malformed.

On line 5, there is a trailing space. It seems like if this is removed before formatting, the output isn't malformed. So maybe that's relevant.

Input (may need to grab from pastebin https://pastebin.com/raw/F0ymMpWR, I had issues reproducting when copying it from GitHub directly)
```js
describe('strformat', () => {
  test('Those symbols break oxfmt plugin completely:', () => {
    expect(strformat('[{x}]', true, { x: 'a b' })).toBe('[a\\u2028b]')
    expect(strformat('[{x}]', true, { x: 'a b' })).toBe('[a\\u2029b]')
  }) 

  test('xyz', () => {})
})
```

Output
```js
describe('strformat', () => {
  test('Those symbols break oxfmt plugin completely:', () => {
    expect(strformat('[{x}]', true, { x: 'a b' })).toBe('[a\\u2028b]')
    expect(strformat('[{x}]', true, { x: 'a b' })).toBe('[a\\u2029b]')
  }) 

  te

  test('xyz', () => {})
})
```

VS Code output for opening the file and formatting it.
<details>

```
2026-06-11 22:26:57.630 [info] [Trace - 10:26:57 PM] Sending notification 'textDocument/didOpen'.
2026-06-11 22:26:57.630 [info] Params: {
    "textDocument": {
        "uri": "file:///Users/nrayburn/Documents/github/oxc-project/reproductions/oxc-intellij-plugin-502/test.js",
        "languageId": "javascript",
        "version": 238,
        "text": "describe('strformat', () => {\n  test('Those symbols break oxfmt plugin completely:', () => {\n    expect(strformat('[{x}]', true, { x: 'a b' })).toBe('[a\\\\u2028b]')\n    expect(strformat('[{x}]', true, { x: 'a b' })).toBe('[a\\\\u2029b]')\n  }) \n\n  test('xyz', () => {})\n})"
    }
}


2026-06-11 22:27:00.746 [info] [Trace - 10:27:00 PM] Sending request 'textDocument/formatting - (46)'.
2026-06-11 22:27:00.746 [info] Params: {
    "textDocument": {
        "uri": "file:///Users/nrayburn/Documents/github/oxc-project/reproductions/oxc-intellij-plugin-502/test.js"
    },
    "options": {
        "tabSize": 2,
        "insertSpaces": true
    }
}


2026-06-11 22:27:00.747 [info] [Trace - 10:27:00 PM] Received response 'textDocument/formatting - (46)' in 1ms.
2026-06-11 22:27:00.747 [info] Result: [
    {
        "range": {
            "start": {
                "line": 6,
                "character": 4
            },
            "end": {
                "line": 9,
                "character": 2
            }
        },
        "newText": "\n\n  test('xyz', () => {})\n})\n"
    }
]


2026-06-11 22:27:01.003 [info] [Trace - 10:27:01 PM] Sending notification 'textDocument/didChange'.
2026-06-11 22:27:01.003 [info] Params: {
    "textDocument": {
        "uri": "file:///Users/nrayburn/Documents/github/oxc-project/reproductions/oxc-intellij-plugin-502/test.js",
        "version": 239
    },
    "contentChanges": [
        {
            "text": "describe('strformat', () => {\n  test('Those symbols break oxfmt plugin completely:', () => {\n    expect(strformat('[{x}]', true, { x: 'a b' })).toBe('[a\\\\u2028b]')\n    expect(strformat('[{x}]', true, { x: 'a b' })).toBe('[a\\\\u2029b]')\n  }) \n\n  te\n\n  test('xyz', () => {})\n})\n"
        }
    ]
}
```

</details>

FYI @leaysgur @Sysix

**Comment by leaysgur** [comment]
Investigated w/ Claude.

The formatting output itself is correct at the byte level (hence CLI/playground are fine).
The bug is in the byte offset → LSP position conversion.

- The language server converts edit offsets to (line, character) via `get_line_column` in `oxc_data_structures::rope`
  - backed by `ropey` (default `unicode_lines` feature), which counts LS (U+2028) / PS (U+2029) as line breaks
- The LSP spec defines line breaks as CR / LF / CRLF only
  - so with one LS and one PS before the edit position, the reported range is shifted down by 2 lines, producing the corrupted output
- `apps/oxlint/src/lsp/` uses the same helpers, so diagnostic positions are likely shifted the same way in files containing LS/PS

Note for the fix: these helpers have two consumers with opposite requirements.

`oxc_transformer` (JSX `__source`) needs the JS definition where LS/PS are valid line breaks (current behavior is correct there), while LSP needs its own. Disabling ropey's `unicode_lines` would break the transformer via feature unification, so the LSP side needs a separate LSP-compliant conversion.
